README.md
NEC BaaS Custom API for Enebular InfoMotion
===========================================

概要
----

株式会社 ウフル が提供する [Enebular](https://enebular.com) [InfoMotion](https://docs.enebular.com/ja/InfoMotion/Introduction.html) から NEC モバイルバックエンド基盤(NEC BaaS) に REST API でアクセスしてデータを取得するための CloudFunctions カスタム API です。
本 API を利用することで、 NEC BaaS に蓄積した IoT データなどの JSON オブジェクトデータを、 InfoMotion を使って可視化することができます。

以下にシステム構成図を示します。

<div align="center">
<img src="https://github.com/nec-baas/baas-infomotion/blob/master/img/baas-infomotion_%E3%82%B7%E3%82%B9%E3%83%86%E3%83%A0%E6%A7%8B%E6%88%90%E5%9B%B3.jpg?raw=true" width="80%" alt="システム構成図">
</div>

# カスタム API 登録
NEC BaaS から JSON オブジェクトデータを取得するカスタム API を登録します。
以降の説明でテナント、アプリケーション、ユーザの入力を行うため、事前に登録してください。
また、テナント情報の CORS 設定を有効にし、 CORS 許可ドメインに InfoMotion のドメインを追加してください。

## API 定義登録
本プロジェクトに格納されている API 定義を登録します。
api.yaml の内容をコピーし、デベロッパーコンソールの API Gateway から登録します。
Basic 認証を行う場合は、 x-acl の値を実行許可するユーザIDやグループ名に設定することを推奨します。以下に例を示します。

    x-acl:
       - 1234567890abcdefxxxxxxxx  # 特定ユーザを指定する場合
       - g:infomotion              # infomotion グループに所属するユーザを指定する場合
       - g:authenticated           # 認証ユーザ全員を指定する場合

ACL の詳細は、 [NEC BaaS のマニュアル](https://nec-baas.github.io/baas-manual/latest/developer/ja/developer/functions/acl.html) をご参照ください。

## Function 定義登録
本プロジェクトに格納されている Function 定義を登録します。
function.yaml の内容をコピーし、デベロッパーコンソールの Cloud Functions > Functions から登録します。登録する際の名前は function.yaml の1行目に記載されている Function 名を入力してください。

## ユーザコード登録

### ユーザコードの取得
ユーザコードは、コンパイル済みバイナリを [Releases](https://github.com/nec-baas/baas-infomotion/releases) より取得し、登録してください。

#### ビルド
ユーザコードをソースからビルドすることも可能です。
以下のコマンドでビルドします。

    $ mvn verify

ビルドが完了すると target フォルダ配下に 「infomotion-7.5.0-cloudfn.tar.gz」 が作成されます。

### バケット作成
デベロッパーコンソールのファイルバケットからバケット名を CUSTOM_CODE としてファイルバケットを作成します。
ACL の設定は、バケット ACL 、コンテンツ ACL 共に全ての権限を空に設定してください。

### ファイル登録
ファイルバケット一覧から作成したファイルバケットを選択し、 infomotion-7.5.0-cloudfn.tar.gz を登録してください。

# アクセス方法
Enebular InfoMotion からの NEC BaaS カスタムAPI の呼び出し方法について説明します。
カスタムAPI の呼び出しには以下の設定を行います。

* InfoType
* DataSource
* InfoMotion

## InfoType 設定
infomotion-tool を使用して作成、もしくは [Discover Assets](https://enebular.com/discover) から任意の InfoType を選択します。
ここでは "Discover Assets" に登録されている Enebular が提供する linechart を使用して説明します。

## DataSource 設定
Enebular のプロジェクトダッシュボードから DataSource を追加します。

DataSource 設定画面では以下の項目が設定できます。下記項目に値を設定します。

|設定項目             |説明                                                                             |
|:--------------------|:--------------------------------------------------------------------------------|
|Title                |任意のタイトル                                                                   |
|DataSource Type      |NEC-BaaS                                                                         |
|Endpoint             |https://{APIサーバホスト名}/api/1/{テナント名}/api/{API名}/{サブパス}?{検索条件} |
|App Id               |NEC BaaS のアプリケーションID                                                     |
|App Key              |NEC BaaS のアプリケーションキー                                                   |
|Username (optional)  |NEC BaaS に登録済みのユーザ名                                                     |
|Password (optional)  |NEC BaaS に登録済みユーザのパスワード                                             |


* テナント名(ID)、アプリケーション ID、アプリケーションキーは NEC BaaS のデベロッパーコンソール上で確認してください。
* API 名は "infomotion" を指定してください。
* サブパスには "/search/{バケット名}" を指定してください。バケット名は参照するオブジェクトバケット名を指定してください。
* 検索条件の設定は任意となります。オブジェクトクエリの検索条件と同フォーマットで where が指定できます。
* DataSource Type は NEC BaaS との連携を行うため、"NEC-BaaS" を指定してください。

## InfoMotion 設定
Enebular のプロジェクトダッシュボードから InfoMotion を追加します。

追加した InfoMotion を開き、 Manage Graphs をクリックします。
右側にグラフを追加するためのメニューが表示されるので Create Graph をクリックします。


下記項目に値を設定し、 Add をクリックしてグラフを追加してください。

|設定項目               |説明                               |
|:----------------------|:----------------------------------|
|Name                   |任意の名前                         |
|Select InfoType        |登録した InfoType                   |
|Select Data DataSource |登録した DataSource                 |
|label                  |凡例                               |
|value                  |値のキー名                         |
|limit                  |表示件数                           |


ChangeLog.txt
2018-12-25: ver 7.5.0

    * 正式版リリース

2018-11-29: ver 7.5.0 beta2

    * 改善: Enebular Infomotion Data Source Schema の新スキーマに対応

2018-11-22: ver 7.5.0 beta1

    * 初版


src/test/java/com/nec/baas/cloudfn/InfoMotionTest.java
package com.nec.baas.cloudfn;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.*;
import static org.mockito.Mockito.*;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.stubbing.Answer;
import org.mockito.stubbing.Stubber;
import org.slf4j.Logger;

import com.nec.baas.cloudfn.sdk.ApigwResponse;
import com.nec.baas.cloudfn.sdk.ClientContext;
import com.nec.baas.cloudfn.sdk.Context;
import com.nec.baas.core.NbBucketMode;
import com.nec.baas.core.NbCallback;
import com.nec.baas.core.NbErrorInfo;
import com.nec.baas.core.NbListCallback;
import com.nec.baas.core.NbRestExecutor;
import com.nec.baas.core.NbService;
import com.nec.baas.core.internal.NbGenericSessionToken;
import com.nec.baas.core.internal.NbRestExecutorFactory;
import com.nec.baas.core.internal.NbServiceImpl;
import com.nec.baas.generic.internal.NbGenericRestExecutor;
import com.nec.baas.http.NbHttpClient;
import com.nec.baas.json.NbJSONObject;
import com.nec.baas.object.NbObject;
import com.nec.baas.object.internal.NbObjectBucketImpl;
import com.nec.baas.object.internal.NbObjectBucketManagerImpl;

class InfoMotionTest {

    @InjectMocks
    private InfoMotion infomotion;

    @Mock
    Context context;

    @Mock
    Logger logger;

    @Mock
    ClientContext clientContext;

    private static final String BUCKET_NAME = "test_bucket";

    private NbService service = new NbServiceImpl();

    @BeforeEach
    public void before() {
        MockitoAnnotations.initMocks(this);

        when(context.clientContext()).thenReturn(clientContext);
        when(context.logger()).thenReturn(logger);
        NbServiceImpl.__clearSingleton();
        ((NbServiceImpl) service).initialize("tenantid123", "appId123", "appKey123", "http://unittest/api/1/tenantid/objects/test_bucket",
                new NbGenericSessionToken(), new GenericRestExecutorFactory());
    }

    /**
     * クエリパラメータを使用し、クエリが実行できること<br>
     * レスポンスがDataSourceスキーマに沿った形式で取得できること<br>
     * クエリパラメータ：start指定<br>
     * パスパラメータ：test_bucket
     */
    @Test
    public void testQueryParameterStart() {

        Date baseDate = new Date();
        Long start = baseDate.getTime() - 60 * 60 * 1000;

        when(clientContext.queryParams()).thenReturn(createQueryParams(start, null, null));
        when(clientContext.pathParams()).thenReturn(createPathParams(BUCKET_NAME));
        executeQuery(ApigwResponse.ok().entity(createApigwResponseData().toString()).build(), createSuccessAnswer());
    }

    /**
     * クエリパラメータを使用し、クエリが実行できること<br>
     * レスポンスがDataSourceスキーマに沿った形式で取得できること<br>
     * クエリパラメータ：end指定<br>
     * パスパラメータ：test_bucket
     */
    @Test
    public void testQueryParameterEnd() {

        Date baseDate = new Date();
        Long end = baseDate.getTime();

        when(clientContext.queryParams()).thenReturn(createQueryParams(null, end, null));
        when(clientContext.pathParams()).thenReturn(createPathParams(BUCKET_NAME));
        executeQuery(ApigwResponse.ok().entity(createApigwResponseData().toString()).build(), createSuccessAnswer());
    }

    /**
     * クエリパラメータを使用し、クエリが実行できること<br>
     * レスポンスがDataSourceスキーマに沿った形式で取得できること<br>
     * クエリパラメータ：start, end指定<br>
     * パスパラメータ：test_bucket
     */
    @Test
    public void testQueryParameterStartEnd() {

        Date baseDate = new Date();
        Long start = baseDate.getTime() - 60 * 60 * 1000;
        Long end = baseDate.getTime();

        when(clientContext.queryParams()).thenReturn(createQueryParams(start, end, null));
        when(clientContext.pathParams()).thenReturn(createPathParams(BUCKET_NAME));
        executeQuery(ApigwResponse.ok().entity(createApigwResponseData().toString()).build(), createSuccessAnswer());
    }

    /**
     * クエリパラメータを使用し、クエリが実行できること<br>
     * レスポンスがDataSourceスキーマに沿った形式で取得できること<br>
     * クエリパラメータ：start, end, where指定<br>
     * パスパラメータ：test_bucket
     */
    @Test
    public void testQueryParameterStartEndWhere() {

        Date baseDate = new Date();
        Long start = baseDate.getTime() - 60 * 60 * 1000;
        Long end = baseDate.getTime();

        when(clientContext.queryParams()).thenReturn(createQueryParams(start, end, "{\"name\":\"categoryA\"}"));
        when(clientContext.pathParams()).thenReturn(createPathParams(BUCKET_NAME));
        executeQuery(ApigwResponse.ok().entity(createApigwResponseData().toString()).build(), createSuccessAnswer());
    }

    /**
     * クエリ結果に"ts"キーが含まれている場合は"_ts"キーに格納されること
     */
    @Test
    public void testTsKeyDuplicated() {
        when(clientContext.queryParams()).thenReturn(createQueryParams(null, null, null));
        when(clientContext.pathParams()).thenReturn(createPathParams(BUCKET_NAME));
        executeQuery(ApigwResponse.ok().entity(createApigwResponseDataTsKey().toString()).build(), createSuccessAnswerTsKeyExist());

    }

    /**
     * "_ts"キーが重複した場合は"ts"キーデータで上書きされること
     */
    @Test
    public void testUnderscoreTsKeyDuplicated() {
        when(clientContext.queryParams()).thenReturn(createQueryParams(null, null, null));
        when(clientContext.pathParams()).thenReturn(createPathParams(BUCKET_NAME));
        executeQuery(ApigwResponse.ok().entity(createApigwResponseDataTsKey().toString()).build(), createSuccessAnswerUnderscoreTsValueOverwrite());

    }

    /**
     * クエリパラメータのstart, endに数値以外を設定し、400エラーとなること
     */
    @Test
    public void testInvalidQueryParameter() {
        when(clientContext.queryParams()).thenReturn(createQueryParams("start", "end", "{\"name\":\"categoryA\"}"));
        when(clientContext.pathParams()).thenReturn(createPathParams(BUCKET_NAME));
        NbJSONObject msg = new NbJSONObject();
        msg.put("error", "Parameter Error");
        executeQuery(ApigwResponse.status(400).entity(msg).build(), null);
    }

    /**
     * クエリパラメータのwhereにJSON文字列以外の値を設定し、400エラーとなること
     */
    @Test
    public void testInvalidQueryParameterWhere() {

        when(clientContext.queryParams()).thenReturn(createQueryParams(null, null, "\"name\":\"categoryA\""));
        when(clientContext.pathParams()).thenReturn(createPathParams(BUCKET_NAME));
        NbJSONObject msg = new NbJSONObject();
        msg.put("error", "Parameter Error");
        executeQuery(ApigwResponse.status(400).entity(msg).build(), null);
    }

    /**
     * オブジェクトクエリが失敗すること(404エラー)<br>
     * クエリパラメータ：start, end指定<br>
     * パスパラメータ：not_found_bucket
     */
    @Test
    public void testBucketNotFound() {

        when(clientContext.queryParams()).thenReturn(createQueryParams(null, null, "{\"name\":\"categoryA\"}"));
        when(clientContext.pathParams()).thenReturn(createPathParams("notfound"));
        executeQuery(ApigwResponse.status(404).entity("{\"error\":\"No such bucket\"}").build(), createFailureAnswer(404, "{\"error\":\"No such bucket\"}"), "notfound");
    }

    /**
     * オブジェクトクエリ結果の日付パース処理に失敗し、500エラーとなること<br>
     * パスパラメータ：test_bucket
     */
    @Test
    public void testInvalidQueryResult() {

        when(clientContext.queryParams()).thenReturn(createQueryParams(null, null, null));
        when(clientContext.pathParams()).thenReturn(createPathParams(BUCKET_NAME));

        String err = "";
        try {
            SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
            sdf.parse("2018-10-31 06:44:35.946");
        } catch (ParseException e) {
            err = e.getMessage();
        }
        NbJSONObject msg = new NbJSONObject();
        msg.put("error", err);

        executeQuery(ApigwResponse.status(500).entity(msg).build(), createSuccessAnswerInvalidUpdatedAt());
    }

    private Answer<Void> createSuccessAnswer() {
        return invocation -> {

            NbCallback<List<NbObject>> cb = invocation.getArgument(1);
            List<NbObject> ret = createSucceedResponseData();

            cb.onSuccess(ret);
            return null;
        };
    }

    private Answer<Void> createSuccessAnswerInvalidUpdatedAt() {

        return invocation -> {

            NbCallback<List<NbObject>> cb = invocation.getArgument(1);
            List<NbObject> ret = createSucceedResponseDataInvalidUpdatedAt();

            cb.onSuccess(ret);
            return null;
        };
    }

    private Answer<Void> createSuccessAnswerTsKeyExist() {

        return invocation -> {

            NbCallback<List<NbObject>> cb = invocation.getArgument(1);
            List<NbObject> ret = createSucceedResponseDataTsKey();

            cb.onSuccess(ret);
            return null;
        };
    }

    private Answer<Void> createSuccessAnswerUnderscoreTsValueOverwrite() {

        return invocation -> {

            NbCallback<List<NbObject>> cb = invocation.getArgument(1);
            List<NbObject> ret = createSucceedResponseDataUnderscoreTsKey();

            cb.onSuccess(ret);
            return null;
        };
    }

    private Answer<Void> createFailureAnswer(int status, String reason) {
        return invocation -> {

            NbCallback<List<NbObject>> cb = invocation.getArgument(1);

            NbErrorInfo err = new NbErrorInfo(reason);
            cb.onFailure(status, err);
            return null;
        };
    }

    private void executeQuery(ApigwResponse verifyResponse, Answer<Void> ans, String bucketName) {

        try {

            NbObjectBucketImpl bucketImpl = spy(new NbObjectBucketImpl(service, bucketName, NbBucketMode.ONLINE));
            bucketImpl.setDescription("Description");

            NbObjectBucketManagerImpl bucketMngMock = mock(NbObjectBucketManagerImpl.class);
            ((NbServiceImpl) service).setObjectBucketManager(bucketMngMock);

            when(context.nebula()).thenReturn(service);
            doReturn(bucketImpl).when(bucketMngMock).getBucket(bucketName);

            if (ans != null) {
                Stubber stub = doAnswer(ans);
                stub.when(bucketImpl).query(any(), any(NbListCallback.class));
            }

            ArgumentCaptor<ApigwResponse> cap = ArgumentCaptor.forClass(ApigwResponse.class);
            doNothing().when(context).succeed(cap.capture());

            ArgumentCaptor<ApigwResponse> capFail = ArgumentCaptor.forClass(ApigwResponse.class);
            doNothing().when(context).fail(capFail.capture());

            InfoMotion.search(null, context);

            if (verifyResponse.getStatus() >= 200 && verifyResponse.getStatus() <= 299) {
                assertEquals(verifyResponse.getEntity(), cap.getValue().getEntity());
                assertEquals(verifyResponse.getStatus(), cap.getValue().getStatus());
            } else {
                assertEquals(verifyResponse.getEntity(), capFail.getValue().getEntity());
                assertEquals(verifyResponse.getStatus(), capFail.getValue().getStatus());
            }

        } catch (Exception e) {
            e.printStackTrace();
            throw e;
        }
    }

    private void executeQuery(ApigwResponse verifyResponse, Answer<Void> ans) {
        executeQuery(verifyResponse, ans, BUCKET_NAME);
    }

    private static class GenericRestExecutorFactory implements NbRestExecutorFactory {
        private NbHttpClient mHttpClient;

        public GenericRestExecutorFactory() {
            mHttpClient = NbHttpClient.getInstance();
        }

        @Override
        public NbRestExecutor create() {
            return new NbGenericRestExecutor(mHttpClient);
        }
    }

    private Map<String, List<String>> createQueryParams(Object start, Object end, String whereJson) {
        Map<String, List<String>> params = new HashMap<>();
        List<String> startList = new ArrayList<>();

        if (start != null) {
            startList.add(start.toString());
            params.put("start", startList);
        }

        if (end != null) {
            List<String> endList = new ArrayList<>();
            endList.add(end.toString());
            params.put("end", endList);
        }

        if (whereJson != null) {
            List<String> whereList = new ArrayList<>();
            whereList.add(whereJson);
            params.put("where", whereList);
        }

        return params;
    }

    private Map<String, String> createPathParams(String bucketName) {
        Map<String, String> pathParam = new HashMap<>();
        pathParam.put("bucketname", bucketName);

        return pathParam;
    }

    private List<NbObject> createApigwResponseData() {
        List<NbObject> ret = new ArrayList<>();

        NbObject obj1 = new NbObject(BUCKET_NAME);
        NbJSONObject objJson1 = new NbJSONObject();
        long timestamp1 = Long.parseLong("1540935875946");
        objJson1.put("_id", "5bd94f53cb6a3b3739e1e92a");
        objJson1.put("createdAt", "2018-10-31T06:44:35.946Z");
        objJson1.put("updatedAt", "2018-10-31T06:44:35.946Z");
        objJson1.put("category", "A");
        objJson1.put("val", "123");
        objJson1.put("ts", timestamp1);
        obj1._setJsonObject(objJson1);

        NbObject obj2 = new NbObject(BUCKET_NAME);
        NbJSONObject objJson2 = new NbJSONObject();
        long timestamp2 = Long.parseLong("1540935876946");
        objJson2.put("_id", "5bd94f53cb6a3b3739e1e92b");
        objJson2.put("createdAt", "2018-10-31T06:44:36.946Z");
        objJson2.put("updatedAt", "2018-10-31T06:44:36.946Z");
        objJson2.put("category", "A");
        objJson2.put("val", "456");
        objJson2.put("ts", timestamp2);
        obj2._setJsonObject(objJson2);

        ret.add(obj1);
        ret.add(obj2);

        return ret;
    }


    private List<NbObject> createApigwResponseDataTsKey() {
        List<NbObject> ret = new ArrayList<>();

        NbObject obj1 = new NbObject(BUCKET_NAME);
        NbJSONObject objJson1 = new NbJSONObject();
        long timestamp1 = Long.parseLong("1540935875946");
        objJson1.put("_id", "5bd94f53cb6a3b3739e1e92a");
        objJson1.put("createdAt", "2018-10-31T06:44:35.946Z");
        objJson1.put("updatedAt", "2018-10-31T06:44:35.946Z");
        objJson1.put("category", "A");
        objJson1.put("val", "123");
        objJson1.put("ts", timestamp1);
        objJson1.put("_ts", "ts_value1");
        obj1._setJsonObject(objJson1);

        NbObject obj2 = new NbObject(BUCKET_NAME);
        NbJSONObject objJson2 = new NbJSONObject();
        long timestamp2 = Long.parseLong("1540935876946");
        objJson2.put("_id", "5bd94f53cb6a3b3739e1e92b");
        objJson2.put("createdAt", "2018-10-31T06:44:36.946Z");
        objJson2.put("updatedAt", "2018-10-31T06:44:36.946Z");
        objJson2.put("category", "A");
        objJson2.put("val", "456");
        objJson2.put("ts", timestamp2);
        objJson2.put("_ts", "ts_value2");
        obj2._setJsonObject(objJson2);

        ret.add(obj1);
        ret.add(obj2);

        return ret;
    }

    private List<NbObject> createSucceedResponseData() {
        List<NbObject> ret = new ArrayList<>();

        NbObject obj1 = new NbObject(BUCKET_NAME);
        obj1.put("_id", "5bd94f53cb6a3b3739e1e92a");
        obj1.put("category", "A");
        obj1.put("val", "123");
        obj1.setUpdatedTime("2018-10-31T06:44:35.946Z");
        obj1.setCreatedTime("2018-10-31T06:44:35.946Z");

        NbObject obj2 = new NbObject(BUCKET_NAME);
        obj2.put("_id", "5bd94f53cb6a3b3739e1e92b");
        obj2.put("category", "A");
        obj2.put("val", "456");
        obj2.setUpdatedTime("2018-10-31T06:44:36.946Z");
        obj2.setCreatedTime("2018-10-31T06:44:36.946Z");

        ret.add(obj1);
        ret.add(obj2);

        return ret;
    }

    private List<NbObject> createSucceedResponseDataInvalidUpdatedAt() {
        List<NbObject> ret = new ArrayList<>();

        NbObject obj1 = new NbObject(BUCKET_NAME);
        obj1.put("_id", "5bd94f53cb6a3b3739e1e92a");
        obj1.put("category", "A");
        obj1.put("val", "123");
        obj1.setUpdatedTime("2018-10-31 06:44:35.946");
        obj1.setCreatedTime("2018-10-31T06:44:35.946Z");

        NbObject obj2 = new NbObject(BUCKET_NAME);
        obj2.put("_id", "5bd94f53cb6a3b3739e1e92b");
        obj2.put("category", "A");
        obj2.put("val", "456");
        obj2.setUpdatedTime("2018-10-31 06:44:36.946");
        obj2.setCreatedTime("2018-10-31T06:44:36.946Z");

        ret.add(obj1);
        ret.add(obj2);

        return ret;
    }

    private List<NbObject> createSucceedResponseDataTsKey() {
        List<NbObject> ret = new ArrayList<>();

        NbObject obj1 = new NbObject(BUCKET_NAME);
        obj1.put("_id", "5bd94f53cb6a3b3739e1e92a");
        obj1.put("category", "A");
        obj1.put("val", "123");
        obj1.put("ts", "ts_value1");
        obj1.setUpdatedTime("2018-10-31T06:44:35.946Z");
        obj1.setCreatedTime("2018-10-31T06:44:35.946Z");

        NbObject obj2 = new NbObject(BUCKET_NAME);
        obj2.put("_id", "5bd94f53cb6a3b3739e1e92b");
        obj2.put("category", "A");
        obj2.put("val", "456");
        obj2.put("ts", "ts_value2");
        obj2.setUpdatedTime("2018-10-31T06:44:36.946Z");
        obj2.setCreatedTime("2018-10-31T06:44:36.946Z");

        ret.add(obj1);
        ret.add(obj2);

        return ret;
    }

    private List<NbObject> createSucceedResponseDataUnderscoreTsKey() {
        List<NbObject> ret = new ArrayList<>();

        NbObject obj1 = new NbObject(BUCKET_NAME);
        obj1.put("_id", "5bd94f53cb6a3b3739e1e92a");
        obj1.put("category", "A");
        obj1.put("val", "123");
        obj1.put("ts", "ts_value1");
        obj1.put("_ts", "_ts_value1");
        obj1.setUpdatedTime("2018-10-31T06:44:35.946Z");
        obj1.setCreatedTime("2018-10-31T06:44:35.946Z");

        NbObject obj2 = new NbObject(BUCKET_NAME);
        obj2.put("_id", "5bd94f53cb6a3b3739e1e92b");
        obj2.put("category", "A");
        obj2.put("val", "456");
        obj2.put("ts", "ts_value2");
        obj2.put("_ts", "_ts_value2");
        obj2.setUpdatedTime("2018-10-31T06:44:36.946Z");
        obj2.setCreatedTime("2018-10-31T06:44:36.946Z");

        ret.add(obj1);
        ret.add(obj2);

        return ret;
    }
}


src/main/java/com/nec/baas/cloudfn/InfoMotion.java
package com.nec.baas.cloudfn;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Map;

import javax.ws.rs.core.MediaType;

import com.nec.baas.cloudfn.sdk.ApigwResponse;
import com.nec.baas.cloudfn.sdk.Context;
import com.nec.baas.core.NbErrorInfo;
import com.nec.baas.core.NbListCallback;
import com.nec.baas.core.NbService;
import com.nec.baas.json.NbJSONArray;
import com.nec.baas.json.NbJSONObject;
import com.nec.baas.json.NbJSONParser;
import com.nec.baas.object.NbClause;
import com.nec.baas.object.NbObject;
import com.nec.baas.object.NbObjectBucket;
import com.nec.baas.object.NbQuery;
import org.slf4j.Logger;

public class InfoMotion {
    /** QueryParameter : start */
    private static final String QUERY_PARAM_START = "start";
    /** QueryParameter : end */
    private static final String QUERY_PARAM_END = "end";
    /** QueryParameter : where */
    private static final String QUERY_PARAM_WHERE = "where";
    /** PathParameter : bucketname */
    private static final String PATH_PARAM_BUCKETNAME = "bucketname";

    public static void search(Map<String, Object> event, Context context) {
        NbClause clause = new NbClause();
        NbQuery query = new NbQuery();
        String bucketName = context.clientContext().pathParams().get(PATH_PARAM_BUCKETNAME);
        NbService nebula = context.nebula();

        final Logger logger = context.logger();
        logger.debug("headers = " + context.clientContext().headers());
        logger.debug("queryParams = " + context.clientContext().queryParams());
        logger.debug("pathParams = " + context.clientContext().pathParams());

        Map<String, List<String>> queryParams = context.clientContext().queryParams();

        // QueryParams
        query.setClause(clause);

        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
        try {
            if (queryParams.containsKey(QUERY_PARAM_START)) {
                long start = Long.parseLong(queryParams.get(QUERY_PARAM_START).get(0));

                NbClause startClause = new NbClause();
                startClause.greaterThanOrEqual("updatedAt", dateFormat.format(new Date(start)));
                clause.and(startClause);
            }

            if (queryParams.containsKey(QUERY_PARAM_END)) {
                long end = Long.parseLong(queryParams.get(QUERY_PARAM_END).get(0));

                NbClause endClause = new NbClause();
                endClause.lessThanOrEqual("updatedAt", dateFormat.format(new Date(end)));
                clause.and(endClause);
            }
        } catch (NumberFormatException e) {
            ApigwResponse response = createErrorResponse(400, "Parameter Error");
            context.fail(response);
            logger.warn("query parameter error", e);
            return;
        }

        if (queryParams.containsKey(QUERY_PARAM_WHERE)) {
            String whereJson = queryParams.get(QUERY_PARAM_WHERE).get(0);
            NbJSONObject whereJsonObject = NbJSONParser.parse(whereJson);
            if (whereJsonObject == null) {
                ApigwResponse response = createErrorResponse(400, "Parameter Error");
                context.fail(response);
                logger.warn("query parameter error");
                return;

            }
            NbClause where = NbClause.fromJSONObject(whereJsonObject);
            clause.and(where);
        }

        // Projection
        NbJSONObject projection = new NbJSONObject();
        projection.put("etag", 0);
        projection.put("ACL", 0);
        query.setProjection(projection);

        // クエリ実行
        NbObjectBucket bucket = nebula.objectBucketManager().getBucket(bucketName);
        bucket.query(query, new NbListCallback<NbObject>() {
            @Override
            public void onSuccess(List<NbObject> objects) {

                try {
                    context.succeed(createResponse(objects));
                } catch (ParseException e) {
                    ApigwResponse response = createErrorResponse(500, e.getMessage());
                    context.fail(response);
                }
            }

            @Override
            public void onFailure(int statusCode, NbErrorInfo errorInfo) {
                String reasonJson = errorInfo.getReason();
                ApigwResponse response =
                        ApigwResponse.status(statusCode).contentType(MediaType.APPLICATION_JSON).entity(reasonJson).build();
                context.fail(response);
            }
        });
    }

    /**
     * Data Source スキーマに沿ったレスポンスを作成する<br>
     * https://docs.enebular.com/ja/infomotion/DataSourceSchema.html
     * 
     * @param objects クエリ実行結果
     * @return レスポンス
     * @throws ParseException ParseException
     */
    private static ApigwResponse createResponse(List<NbObject> objects) throws ParseException {
        DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");

        NbJSONArray<NbJSONObject> response = new NbJSONArray<>();
        for (NbObject object : objects) {
            NbJSONObject json = new NbJSONObject();

            // value
            json.put("_id", object.getObjectId());
            json.put("createdAt", object.getCreatedTime());
            json.put("updatedAt", object.getUpdatedTime());

            object.forEach(json::put);

            // rename ts
            if (json.containsKey("ts")) {
                json.put("_ts", json.get("ts"));
            }

            // timestamp
            long timestamp = dateFormat.parse(object.getUpdatedTime()).getTime();
            json.put("ts", timestamp);

            response.add(json);
        }

        return ApigwResponse.ok().contentType(MediaType.APPLICATION_JSON).entity(response.toString()).build();
    }

    /**
     * エラーレスポンス(JSON)を作成する
     * @param statusCode ステータスコード
     * @param message エラーメッセージ
     * @return ApigwResponse
     */
    private static ApigwResponse createErrorResponse(int statusCode, String message) {
        NbJSONObject msgJson = new NbJSONObject();
        msgJson.put("error", message);
        return ApigwResponse.status(statusCode).contentType(MediaType.APPLICATION_JSON).entity(msgJson).build();
    }
}
